### PR TITLE
sanctions/list_unnecessary_reindex

### PR DESCRIPTION
**Pull request by mehdi-fs**

### CODE REVIEW

File: lib/Data/Validate/Sanctions.pm [hunk 1]
```diff
@@ -316,16 +318,20 @@ sub get_sanctioned_info {    ## no critic (RequireArgUnpacking)
...
     $self->{_data}                   //= {};
     $self->{_sanctioned_name_tokens} //= {};
     $self->{_token_sanctioned_names} //= {};
 
     if (-e $sanction_file) {
-        return $self->{_data} if stat($sanction_file)->mtime <= $self->{last_time} && $self->{_data};
-        $self->{last_time} = stat($sanction_file)->mtime;
-        $self->{_data}     = LoadFile($sanction_file);
+        return $self->{_data} if stat($sanction_file)->mtime <= $self->{last_modification} && $self->{_data};
+        $self->{last_modification} = stat($sanction_file)->mtime;
```
Comment (kirill-binary) [inline lib/Data/Validate/Sanctions.pm:329]: the `stat` requires reading from the file system and it's not a free operation, I'd set some threshold when we even don't bother to check modification time. for example, if less than  5-10 min passed from the last  check, do nothing.

File: lib/Data/Validate/Sanctions.pm [hunk 2]
```diff
@@ -316,16 +318,20 @@ sub get_sanctioned_info {    ## no critic (RequireArgUnpacking)
...
+        $self->{_data}             = LoadFile($sanction_file);
     }
+
+    return $self->{_data} if $self->{last_modification} <= $self->{last_index};
```
Comment (kirill-binary) [inline lib/Data/Validate/Sanctions.pm:333]: Do we need this line? if data wasn't modified we return at #328, if we load new file then we always need to update the index... What case are we handling here?

File: lib/Data/Validate/Sanctions/Redis.pm [hunk 3]
```diff
@@ -50,12 +53,13 @@ sub get_sanctioned_info {
...
 
-    $self->{last_time}               //= 0;
+    $self->{last_modification}       //= 0;
+    $self->{last_index}              //= 0;
     $self->{_data}                   //= {};
     $self->{_sanctioned_name_tokens} //= {};
     $self->{_token_sanctioned_names} //= {};
 
-    my $last_time = $self->{last_time};
+    my $last_modification = $self->{last_modification};
```
Comment (kirill-binary) [inline lib/Data/Validate/Sanctions/Redis.pm:62]: same here let's check when the last time we checked for an update if it's less than 5-10 min. no point in poking Redis, in the current implementation we'll fetch from Redis whore sanction list every dozen ms

File: lib/Data/Validate/Sanctions/Redis.pm [hunk 4]
```diff
@@ -69,15 +73,17 @@ sub _load_data {
             $self->{_data}->{$source}->{verified} = $verified // 0;
             $self->{_data}->{$source}->{updated}  = $updated;
             $self->{_data}->{$source}->{error}    = $error // '';
-            $last_time                            = $updated if $updated > $last_time;
+            $last_modification                    = $updated if $updated > $last_modification;
```
Comment (kirill-binary) [inline lib/Data/Validate/Sanctions/Redis.pm:76]: Lets break down this. 
```
my ($content, $verified, $updated, $error) = $self->{connection}->hmget("SANCTIONS::$source", qw/content verified updated error/)->@*;
```
let's get first only `updated`, if it's not any different from what we have in memory, no point to bother Redis with transferring whole sanctions lists if it's the same as we have  in memore

Comment (mehdi-fs) [reply lib/Data/Validate/Sanctions/Redis.pm:76]: yeah make sense

File: lib/Data/Validate/Sanctions/Redis.pm [hunk 5]
```diff
@@ -50,34 +53,40 @@ sub get_sanctioned_info {
...
 
-    $self->{last_time}               //= 0;
+    $self->{last_modification}       //= 0;
+    $self->{last_index}              //= 0;
     $self->{_data}                   //= {};
     $self->{_sanctioned_name_tokens} //= {};
     $self->{_token_sanctioned_names} //= {};
 
-    my $last_time = $self->{last_time};
+    return $self->{_data} if $self->{_data} and $self->{last_modification} + $self->IGNORE_OPERATION_INTERVAL > time;
```
Comment (kirill-binary) [inline lib/Data/Validate/Sanctions/Redis.pm:62]: Here we should to compare not with last modification time but with time when we last time loaded file into memory... we update file once a day... optimization will work only for first 8 minutes after file was updated...

**Pull request merged**